Three APIs for finding all and a single app and adding a new app

titay2 opened the pull request:
@gr2m @flyjwayur  Conflicts on branch 32 has been solved on this branch

titay2 committed:
three APIs for finding all and a single app and adding a new app

gr2m commented:
I get an error when I try to create a new app:

> main.js:38 Uncaught ReferenceError: applist is not defined

I’m looking into it

titay2 commented:
@gr2m  its that I forgot to import  the indexAPI file on the index.html page

gr2m committed:
fix: load `scripts/indexApi.js` and remove `styles/simple-sidebar.css`

gr2m commented:
@titay2 I’ve updated the pull request, make sure to pull the latest changes before you continue to work on it. Now I see both, the list of apps and the form, shall I look into it or do you want to have a look?

<img width="792" alt="screen shot 2016-08-18 at 11 43 41" src="https://cloud.githubusercontent.com/assets/39992/17786316/14b3c94c-6539-11e6-8bb4-86b22e428a82.png">

titay2 commented:
@gr2m  I had made a different issue (issue #32 )where everything was working and up to date but it had some merging conflict so i manually made this branch and am still confused what is happening

gr2m commented:
I’m looking into it

gr2m committed:
fix: #container2 should be #detail_app_page

gr2m commented:
I think I found the issue and resolved it via https://github.com/Rookies-RGSOC2016/hoodie-electron-app/pull/35/commits/b8697947edd98bf8c638e24e2364a02d7626a9a7 – does that make sense?

Next issue I see is that when you hit cancel, both are shown again, I’m having a look

titay2 commented:
yes it does :)

gr2m committed:
fix: hide create app form on cancel

gr2m commented:
I fixed it via https://github.com/Rookies-RGSOC2016/hoodie-electron-app/pull/35/commits/a085edc29a1e49c6e1742c62e0dbe2df2eb8889d

gr2m commented:
when I create an app, the list does not show up :D looking into it

gr2m committed:
fix: Promise usage

@titay2 make sure to review and understand these changes to the promises, why I added the `.then` callbacks. Please let me know if you have any questions

gr2m commented:
fixed via https://github.com/Rookies-RGSOC2016/hoodie-electron-app/pull/35/commits/6f40f3b6dd8ee3273869643f6e8aef05b7eff5b9, make sure to review it @titay2 @flyjwayur let me know if you have any questions

gr2m commented on scripts/main.js:
```diff
@@ -1,32 +1,37 @@
-		  name : appname
-		})
-	makeAppList();
-	showAppDetail(app.id);
```
the problem here was that app.id was undefined, because `applist.add()` returns a promise. So you have to add a `.then()` callback which gets an `app` argument which then has the `app.id` defined

gr2m commented on scripts/main.js:
```diff
@@ -1,32 +1,37 @@
+}
 });
 function showAppDetail(id) {
-  var app = applist.find(id);
```
similar problem as above. `app` here is a promise, because `applist.find()` returns a promise, not the actual app object. I know that is a bit complex, but unfortunately we have to make these APIs asynchronous as we will write to files and/or databases in future, and we cannot make this a synchronous process that blocks the execution of other JavaScript. This is why we use Promises

gr2m committed:
fix: use `app.id` for the hash in the URL, as the name can contain non-characters like spaces, while `app.id` is save to use in URLs

titay2 commented:
Thank you so much for the help @gr2m I have lost the concept that promises are not like methods in a class they cant be executed just on a call, they just return promises

gr2m commented:
okay great, I’ll start my code review from here in another pull request, ok? You are done with work for the weekend, right?

titay2 commented:
I think so :)
